### ISSUE TITLE
train dVAE lr scheduler step in wrong place?

### ISSUE THREAD

**Issue by yangyixiao** [lucidrains/DALLE-pytorch#278]
Hi! In train_vae.py line 266
`            if not using_deepspeed:
                    # Scheduler is automatically progressed after the step
                    # when using DeepSpeed.
                    distr_sched.step()
`
 I found the sheduler was stepped. But in my understanding, learning rate should be updated at the end of each epoch. Otherwise, the learning rate becomes 0 in the first epoch in my experiment.

Is this a bug or my misunderstanding? Thanks for help

**Comment by janEbert** [comment]
Hey Yangyi! Great catch, thank you so much!

The solution, quoted from the DeepSpeed [devdocs](https://deepspeed.readthedocs.io/en/latest/schedulers.html):

> When using a DeepSpeed’s learning rate scheduler (specified in the ds_config.json file), DeepSpeed calls the step() method of the scheduler at every training step (when model_engine.step() is executed). When not using a DeepSpeed’s learning rate scheduler:
>
> - if the schedule is supposed to execute at every training step, then the user can pass the scheduler to deepspeed.initialize when initializing the DeepSpeed engine and let DeepSpeed manage it for update or save/restore.
> - if the schedule is supposed to execute at any other interval (e.g., training epochs), then the user should NOT pass the scheduler to DeepSpeed during initialization and must manage it explicitly.

So we must _not_ pass in the LR scheduler into the `distr_backend.distribute` method (instead give `None`).
Also, we'd remove the if-guard around the scheduler step so it's executed even with DeepSpeed.
Do you want to PR the fix?

We should probably also do the same in `train_dalle.py` to avoid future surprises.
